src/eos-companion-app-integration-helper.c

```diff
@@ -456,6 +457,41 @@ load_desktop_info_key_file_for_app_id (const gchar      *app_id,
   return g_steal_pointer (&system_desktop_app_info);
 }
 
+static void
+lookup_eknservices_version (const gchar        *runtime_version,
+                            const gchar const **out_eknservices_name,
+                            const gchar const **out_search_provider_name)
+{
+  g_return_if_fail (out_eknservices_name != NULL);
+  g_return_if_fail (out_search_provider_name != NULL);
+
+  static const struct {
+    const gchar *runtime_version;
+    const gchar *eknservices_name;
+    const gchar *search_provider_name;
+  } runtime_to_eknservices_versions[] = {
+    { "2", "EknServices2", "SearchProviderV2" },
+    { "3", "EknServices2", "SearchProviderV2" },
+    { "4", "EknServices3", "SearchProviderV3" }
+  };
+  static const gsize n_runtime_to_eknservices_version = G_N_ELEMENTS (runtime_to_eknservices_versions);
+  gsize i = 0;
+
+  for (; i < n_runtime_to_eknservices_version; ++i)
+    {
+      if (g_strcmp0 (runtime_version,
+                     runtime_to_eknservices_versions[i].runtime_version) == 0)
+        {
+          *out_eknservices_name = runtime_to_eknservices_versions[i].eknservices_name;
+          *out_search_provider_name = runtime_to_eknservices_versions[i].search_provider_name;
+          return;
+        }
+    }
+
+  *out_eknservices_name = NULL;
+  *out_search_provider_name = NULL;
```

Comment:
Three things about this function:
- it's not typical to see functions with out params that return void, you should probably return a boolean that indicates whether this succeeded
- it's not typical for functions to touch out params in case of failure (other than a gerror param)
- by not copying the strings in the out params, you're extending the scope of const strings outside of where they're declared. Even if it may work, I believe that's undefined behavior and I had to spend a ton of time debugging an issue last week that was caused just by that: https://git.gnome.org/browse/sushi/commit/?id=0e7856afb9f2194579cfc0e5545f2a402df6f020. Please copy those strings :-)

Reply:
I'll make the function return a boolean.

As for the string lifetime, that *should* work because the strings have static lifetime. However, I can also just copy them at the cost of a negligible amount of memory usage